modules/core/src/main/scala/com.snowplowanalytics/iglu.schemaddl/parquet/Field.scala

```diff
@@ -0,0 +1,157 @@
+      case other => other
+    }
+    field.copy(name = normalizeName(field), fieldType = fieldType)
+  }
+
+  private[parquet] def normalizeName(field: Field): String =
+    StringUtils.snakeCase
+      .andThen(replaceDisallowedCharacters)(field.name)
+
+  /**
+   * Replaces disallowed parquet column characters with underscore
+   */
+  private val replaceDisallowedCharacters: String => String = str =>
+    str.replaceAll("[^A-Za-z0-9_]", "_")
+
+  private[parquet] final case class NullableType(value: Type, nullability: JsonNullability)
+
+  /* An interim nullability type, which should not be public beyond schema-ddl.
+   *
+   * It represents whether the JsonSchema "type" field allows explicit nulls. Semantically it has a
+   * different meaning to the parquet [[Type.Nullability]], which is also concerned with missing values.
+   */
+  private[parquet] sealed trait JsonNullability
+
+  private[parquet] object JsonNullability {
+    case object ExplicitlyNullable extends JsonNullability
+    case object NoExplicitNull extends JsonNullability
+
+    def extractFrom(`type`: CommonProperties.Type): JsonNullability = {
+      if (`type`.nullable) {
+        JsonNullability.ExplicitlyNullable
+      } else
+        JsonNullability.NoExplicitNull
+    }
+  }
+
+  private[parquet] def buildType(topSchema: Schema): NullableType = {
+    topSchema.`type` match {
+      case Some(types) if types.possiblyWithNull(CommonProperties.Type.Object) =>
+        NullableType(
+          value = buildObjectType(topSchema), 
+          nullability = JsonNullability.extractFrom(types)
+        )
+
+      case Some(types) if types.possiblyWithNull(CommonProperties.Type.Array) =>
+        NullableType( 
+          value = buildArrayType(topSchema), 
+          nullability = JsonNullability.extractFrom(types)
+        )
+
+      case _ =>
+        provideSuggestions(topSchema) match {
+          case Some(matchingSuggestion) => matchingSuggestion
+          case None => jsonType(topSchema)
+        }
+    }
+  }
+
+  private def buildObjectType(topSchema: Schema): Type = {
+    val subfields = topSchema.properties.map(_.value).getOrElse(Map.empty)
+    if (subfields.nonEmpty) {
+      val requiredKeys = topSchema.required.toList.flatMap(_.value)
+      val structFields = subfields
+        .toList
+        .map { case (key, schema) =>
+          val isSubfieldRequired = requiredKeys.contains(key)
+          val constructedType = buildType(schema)
+          val nullability = isFieldNullable(constructedType.nullability, isSubfieldRequired)
+          Field(key, constructedType.value, nullability)
+        }
+        .sortBy(_.name)
+      Type.Struct(structFields)
+    } else {
+      Type.Json
+    }
+  }
+
+  private def buildArrayType(topSchema: Schema): Type.Array = {
+    topSchema.items match {
+      case Some(ArrayProperty.Items.ListItems(schema)) =>
+        val typeOfArrayItem = buildType(schema)
+        val nullability = typeOfArrayItem.nullability match {
```

Comment:
Could be `isFieldNullable` with `enforceItemPresence = true`, but this is also fine